src/ParallelHashMap.h

```diff
@@ -0,0 +1,838 @@
+ *    announce array. First, all locks are set in order to block inserts and
+ *    prevent deadlock. A new table is allocated of twice the size and all
+ *    key/value pairs from the old table, then the pointer is switched to the
+ *    new table and locks are released. Finally the memory needs to be freed.
+ *    To prevent threads currently reading the table from encountering
+ *    segmentation faults, the resizing threads waits for the announce array
+ *    to be free of references to the old table before freeing the memory.
+ */
+template <class K, class V>
+void ParallelHashMap<K,V>::resize()
+{
+  /* acquire all locks in order */
+  #ifdef OPENMP
+  for(size_t i=0; i<_num_locks; i++)
+    omp_set_lock(&_locks[i]);
+  #endif 
+
+  /* recheck if resize needed */
+  if(2*_table->size() < _table->bucket_count())
+  {
+    /* release locks */
+    #ifdef OPENMP
+    for(size_t i=0; i<_num_locks; i++)
+      omp_unset_lock(&_locks[i]);
+    #endif 
+
+    return;
+  }
+
+  /* allocate new hash map of double the size */
+  fixed_hash_map<K,V> *new_map = 
+    new fixed_hash_map<K,V>(2*_table->bucket_count());
+
+  /* get keys, values, and number of elements */
+  K *key_list = _table->keys();
+  V *value_list = _table->values();
+
+  /* insert key/value pairs into new hash map */
+  for(size_t i=0; i<_table->size(); i++)
+    new_map->insert(key_list[i], value_list[i]);
+
+  /* save pointer of old table */
+  fixed_hash_map<K,V> *old_table = _table;
+
+  /* reassign pointer */
+  _table = new_map;
+
+  /* release all locks */
+  #ifdef OPENMP
+  for(size_t i=0; i<_num_locks; i++)
+    omp_unset_lock(&_locks[i]);
+  #endif
+
+  /* delete key and value list */
+  delete[] key_list;
+  delete[] value_list;
+
+  /* wait for all threads to stop reading from the old table */
+  for(size_t i=0; i<_num_threads; i++)
+    while(_announce[i].value == old_table) {};
+
+  /* free memory associated with old table */
+  delete old_table;
+}
+
+/**
+ * @brief Returns the number of key/value pairs in the underlying table
+ * @return number of key/value pairs in the map
+ */
+template <class K, class V>
+size_t ParallelHashMap<K,V>::size()
+{
+  return _table->size();
+}
+
+/**
+ * @brief Returns the number of buckets in the underlying table
+ * @return number of buckets in the map
+ */
+template <class K, class V>
+size_t ParallelHashMap<K,V>::bucket_count()
+{
+  return _table->bucket_count();
+}
+
+/**
+ * @brief Returns the number of locks in the parallel hash map
+ * @return number of locks in the map
+ */
+template <class K, class V>
+size_t ParallelHashMap<K,V>::num_locks()
+{
+  return _num_locks;
+}
+
+/**
+ * @brief Returns an array of the keys in the underlying table
+ * @details All buckets are scanned in order to form a list of all keys
+ *      present in the table and then the list is returned. Threads
+ *      announce their presence to ensure table memory is not freed
+ *      during access.
```

Comment:
Add a NOTE or WARNING here in the ``@details`` comment block that it is the caller's responsibility to free the memory allocated for the returned ``keys`` array.